File: ui/public/acp-client.js
(function (global) {
  const ACP_CLIENT_INFO = {
    name: 'spritz-ui',
    title: 'Spritz ACP UI',
    version: '1.0.0',
  };

  function extractACPText(value) {
    if (value === null || value === undefined) return '';
    if (typeof value === 'string') return value;
    if (Array.isArray(value)) {
      return value.map((item) => extractACPText(item)).filter(Boolean).join('\n');
    }
    if (typeof value !== 'object') return String(value);
    if (typeof value.text === 'string') return value.text;
    if (value.type === 'content' && value.content) return extractACPText(value.content);
    if (value.content) return extractACPText(value.content);
    if (value.resource) {
      if (typeof value.resource.text === 'string') return value.resource.text;
      if (typeof value.resource.uri === 'string') return value.resource.uri;
    }
    return '';
  }

  function createACPClient(options) {
    const {
      wsUrl,
      conversation,
      onStatus,
      onReadyChange,
      onAgentInfo,
      onUpdate,
      onPermissionRequest,
      onPromptStateChange,
      onClose,
      onProtocolError,
    } = options;

    let ws = null;
    let nextId = 1;
    let ready = false;
    let disposed = false;
    const pending = new Map();
    let sessionId = conversation?.spec?.sessionId || '';
    let loadSessionSupported = false;
    let bootstrapComplete = false;

    function createACPError(code, message) {
      const error = new Error(message);
      error.code = code;
      return error;
    }

    function createRPCError(payload) {
      const baseMessage = String(payload?.message || 'ACP request failed.');
      const details =
        typeof payload?.data?.details === 'string' && payload.data.details.trim()
          ? payload.data.details.trim()
          : '';
      const error = createACPError('ACP_RPC_ERROR', details ? `${baseMessage}: ${details}` : baseMessage);
      error.rpcError = payload || null;
      return error;
    }

    function cleanupPending(error) {
      const resolvedError =
        error instanceof Error
          ? error
          : createACPError('ACP_REQUEST_CANCELLED', String(error || 'ACP request cancelled.'));
      pending.forEach(({ reject }) => reject(resolvedError));
      pending.clear();
    }

    function sendRaw(payload) {
      if (!ws || ws.readyState !== WebSocket.OPEN) {
        throw new Error('ACP socket is not connected.');
      }
      ws.send(JSON.stringify(payload));
    }

    function requestRPC(method, params) {
      return new Promise((resolve, reject) => {
        const id = nextId++;
        pending.set(String(id), { resolve, reject, method });
        try {
          sendRaw({ jsonrpc: '2.0', id, method, params });
        } catch (err) {
          pending.delete(String(id));
          reject(err);
        }
      });
    }

    function respond(id, result) {
      sendRaw({ jsonrpc: '2.0', id, result });
    }

    function respondError(id, code, message) {
      sendRaw({ jsonrpc: '2.0', id, error: { code, message } });
    }

    function notify(method, params) {
      sendRaw({ jsonrpc: '2.0', method, params });
    }

    function isMissingSessionError(error) {
      const message = String(error?.message || '').toLowerCase();
      return message.includes('session') && message.includes('not found');
    }

    async function bootstrapSession() {
      onStatus('Negotiating ACP…');
      const init = await requestRPC('initialize', {
        protocolVersion: 1,
        clientCapabilities: {},
        clientInfo: ACP_CLIENT_INFO,
      });
      loadSessionSupported = Boolean(init?.agentCapabilities?.loadSession);
      if (typeof onAgentInfo === 'function') {
        onAgentInfo(init?.agentInfo || null);
      }
      if (!loadSessionSupported) {
        throw createACPError('ACP_LOAD_SESSION_UNSUPPORTED', 'ACP agent does not support session/load.');
      }
      if (!sessionId) {
        throw createACPError('ACP_SESSION_ID_REQUIRED', 'Conversation is missing an ACP session id.');
      }
      onStatus('Loading conversation…');
      try {
        await requestRPC('session/load', {
          sessionId,
          cwd: conversation?.spec?.cwd || '/home/dev',
          mcpServers: [],
        });
      } catch (error) {
        if (isMissingSessionError(error)) {
          const details =
            typeof error?.rpcError?.data?.details === 'string' && error.rpcError.data.details.trim()
              ? error.rpcError.data.details.trim()
              : error.message || 'ACP session not found.';
          throw createACPError('ACP_SESSION_MISSING', details);
        }
        throw error;
      }
      onStatus('Connected');
    }

    function handleIncoming(message) {
      if (message.id !== undefined && (message.result !== undefined || message.error)) {
        const key = String(message.id);
        const pendingRequest = pending.get(key);
        if (!pendingRequest) return;
        pending.delete(key);
        if (message.error) {
          if (pendingRequest.method === 'session/prompt' && typeof onPromptStateChange === 'function') {
            onPromptStateChange(false);
          }
          pendingRequest.reject(createRPCError(message.error));
          return;
        }
        pendingRequest.resolve(message.result);
        if (pendingRequest.method === 'session/prompt' && typeof onPromptStateChange === 'function') {
          onPromptStateChange(false);
        }
        return;
      }

      if (message.method === 'session/update') {
        if (typeof onUpdate === 'function') {
          onUpdate(message.params?.update);
        }
        return;
      }

      if (message.method === 'session/request_permission' && message.id !== undefined) {
        if (typeof onPermissionRequest === 'function') {
          onPermissionRequest({
            params: message.params,
            respond(result) {
              respond(message.id, result);
            },
            reject(messageText) {
              respondError(message.id, -32000, messageText || 'Permission denied.');
            },
          });
        }
        return;
      }

      if (message.id !== undefined) {
        respondError(message.id, -32601, 'Method not supported by Spritz ACP UI.');
      }
    }

    function start() {
      disposed = false;
      ready = false;
      bootstrapComplete = false;
      if (typeof onReadyChange === 'function') {
        onReadyChange(false);
      }
      return new Promise((resolve, reject) => {
        ws = new WebSocket(wsUrl);
        ws.onopen = async () => {
          try {
            await bootstrapSession();
            bootstrapComplete = true;
            ready = true;
            if (typeof onReadyChange === 'function') {
              onReadyChange(true);
            }
            resolve();
          } catch (err) {
            bootstrapComplete = false;
            ready = false;
            if (typeof onReadyChange === 'function') {
              onReadyChange(false);
            }
            reject(err);
            try {
              ws.close();
            } catch {
              // ignore
            }
          }
        };
        ws.onmessage = (event) => {
          try {
            const data = typeof event.data === 'string' ? event.data : new TextDecoder().decode(event.data);
            handleIncoming(JSON.parse(data));
          } catch (err) {
            if (typeof onProtocolError === 'function') {
              onProtocolError(err);
            }
          }
        };
        ws.onerror = () => {
          if (!ready) {
            reject(new Error('Failed to connect to ACP gateway.'));
          }
        };
        ws.onclose = () => {
          ready = false;
          bootstrapComplete = false;
          cleanupPending(createACPError('ACP_CONNECTION_CLOSED', 'ACP connection closed.'));
          if (typeof onReadyChange === 'function') {
            onReadyChange(false);
          }
          if (!disposed && typeof onClose === 'function') {
            onClose('ACP connection closed.');
          }
        };
      });
    }

    return {
      start,
      getConversationId() {
        return conversation?.metadata?.name || '';
      },
      getSessionId() {
        return sessionId || '';
      },
      matchesConversation(targetConversation) {
        const targetConversationId = targetConversation?.metadata?.name || '';
        if (targetConversationId && targetConversationId !== (conversation?.metadata?.name || '')) {
          return false;
        }
        const expectedSessionId = targetConversation?.spec?.sessionId || '';
        if (expectedSessionId && sessionId && expectedSessionId !== sessionId) {
          return false;
        }
        return true;
      },
      isReady() {
        return Boolean(ws && ws.readyState === WebSocket.OPEN && ready && bootstrapComplete);
      },
      async sendPrompt(text) {
        if (!sessionId || !this.isReady()) {
          throw new Error('ACP session is not ready yet.');
        }
        if (typeof onPromptStateChange === 'function') {
          onPromptStateChange(true);
        }
        return requestRPC('session/prompt', {
          sessionId,
          prompt: [{ type: 'text', text }],
        });
      },
      cancelPrompt() {
        if (!sessionId || !this.isReady()) return;
        notify('session/cancel', { sessionId });
        if (typeof onStatus === 'function') {
          onStatus('Cancelling…');
        }
      },
      dispose() {
        disposed = true;
        ready = false;
        bootstrapComplete = false;
        cleanupPending(createACPError('ACP_CLIENT_DISPOSED', 'ACP client disposed.'));
        try {
          if (ws && (ws.readyState === WebSocket.OPEN || ws.readyState === WebSocket.CONNECTING)) {
            ws.close();
          }
        } catch {
          // ignore
        }
      },
    };
  }

  global.SpritzACPClient = {
    createACPClient,
    extractACPText,
  };
})(window);

File: integrations/slack-gateway/acp_client.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"

	"github.com/gorilla/websocket"
)

type acpRPCMessage struct {
	JSONRPC string          `json:"jsonrpc,omitempty"`
	ID      any             `json:"id,omitempty"`
	Method  string          `json:"method,omitempty"`
	Params  json.RawMessage `json:"params,omitempty"`
	Result  json.RawMessage `json:"result,omitempty"`
	Error   *struct {
		Code    int             `json:"code"`
		Message string          `json:"message"`
		Data    json.RawMessage `json:"data,omitempty"`
	} `json:"error,omitempty"`
}

func (g *slackGateway) promptConversation(ctx context.Context, serviceToken, namespace, conversationID, sessionID, cwd, prompt string) (string, bool, error) {
	wsURL, err := g.spritzWebSocketURL("/api/acp/conversations/"+url.PathEscape(conversationID)+"/connect", map[string]string{"namespace": namespace})
	if err != nil {
		return "", false, err
	}
	dialer := websocket.Dialer{HandshakeTimeout: g.cfg.HTTPTimeout}
	headers := http.Header{}
	headers.Set("Authorization", "Bearer "+serviceToken)
	conn, _, err := dialer.DialContext(ctx, wsURL, headers)
	if err != nil {
		return "", false, err
	}
	defer conn.Close()

	client := &acpPromptClient{conn: conn}
	if _, _, err := client.call(ctx, "initialize", map[string]any{
		"protocolVersion":    1,
		"clientCapabilities": map[string]any{},
		"clientInfo": map[string]any{
			"name":    "slack-gateway",
			"title":   "Slack Gateway",
			"version": "1.0.0",
		},
	}, nil); err != nil {
		return "", false, err
	}
	if _, _, err := client.call(ctx, "session/load", map[string]any{
		"sessionId":  sessionID,
		"cwd":        cwd,
		"mcpServers": []any{},
	}, nil); err != nil {
		return "", false, err
	}
	var reply strings.Builder
	if _, promptSent, err := client.call(ctx, "session/prompt", map[string]any{
		"sessionId": sessionID,
		"prompt": []map[string]any{{
			"type": "text",
			"text": prompt,
		}},
	}, func(message *acpRPCMessage) {
		if strings.TrimSpace(message.Method) != "session/update" || len(message.Params) == 0 {
			return
		}
		var payload struct {
			Update map[string]any `json:"update"`
		}
		if err := json.Unmarshal(message.Params, &payload); err != nil {
			return
		}
		if strings.TrimSpace(stringValue(payload.Update["sessionUpdate"])) != "agent_message_chunk" {
			return
		}
		reply.WriteString(extractACPText(payload.Update["content"]))
	}); err != nil {
		return strings.TrimSpace(reply.String()), promptSent, err
	}
	text := strings.TrimSpace(reply.String())
	if text == "" {
		return "", true, fmt.Errorf("agent returned an empty reply")
	}
	return text, true, nil
}

type acpPromptClient struct {
	conn   *websocket.Conn
	nextID int64
}

func (c *acpPromptClient) writeJSON(ctx context.Context, payload any) error {
	if deadline, ok := ctx.Deadline(); ok {
		_ = c.conn.SetWriteDeadline(deadline)
	}
	return c.conn.WriteJSON(payload)
}

func (c *acpPromptClient) respondError(ctx context.Context, id any, code int, message string) error {
	return c.writeJSON(ctx, map[string]any{
		"jsonrpc": "2.0",
		"id":      id,
		"error": map[string]any{
			"code":    code,
			"message": message,
		},
	})
}

func (c *acpPromptClient) call(ctx context.Context, method string, params any, onNotification func(*acpRPCMessage)) (json.RawMessage, bool, error) {
	c.nextID++
	requestID := fmt.Sprintf("rpc-%d", c.nextID)
	delivered := false
	if err := c.writeJSON(ctx, map[string]any{
		"jsonrpc": "2.0",
		"id":      requestID,
		"method":  method,
		"params":  params,
	}); err != nil {
		return nil, false, err
	}
	for {
		if deadline, ok := ctx.Deadline(); ok {
			_ = c.conn.SetReadDeadline(deadline)
		}
		_, payload, err := c.conn.ReadMessage()
		if err != nil {
			return nil, delivered, err
		}
		var message acpRPCMessage
		if err := json.Unmarshal(payload, &message); err != nil {
			return nil, delivered, err
		}
		if message.Method == "session/request_permission" && message.ID != nil {
			delivered = true
			if err := c.respondError(
				ctx,
				message.ID,
				-32000,
				"Permission denied: interactive approvals are unavailable in the Slack gateway.",
			); err != nil {
				return nil, delivered, err
			}
			continue
		}
		if message.Method != "" && message.ID == nil {
			delivered = true
			if onNotification != nil {
				onNotification(&message)
			}
			continue
		}
		if fmt.Sprint(message.ID) != requestID {
			continue
		}
		delivered = true
		if message.Error != nil {
			return nil, delivered, fmt.Errorf("%s", strings.TrimSpace(message.Error.Message))
		}
		return message.Result, delivered, nil
	}
}

func extractACPText(value any) string {
	switch typed := value.(type) {
	case nil:
		return ""
	case string:
		return typed
	case []any:
		parts := make([]string, 0, len(typed))
		for _, item := range typed {
			if text := extractACPText(item); text != "" {
				parts = append(parts, text)
			}
		}
		return strings.Join(parts, "\n")
	case map[string]any:
		if text := stringValue(typed["text"]); text != "" {
			return text
		}
		if content, ok := typed["content"]; ok {
			return extractACPText(content)
		}
		if resource, ok := typed["resource"]; ok {
			return extractACPText(resource)
		}
		if uri := stringValue(typed["uri"]); uri != "" {
			return uri
		}
	}
	return ""
}

func stringValue(value any) string {
	switch typed := value.(type) {
	case string:
		return strings.TrimSpace(typed)
	default:
		return ""
	}
}
